src/task4feedback/simulator/rl/models/simple.py

```diff
@@ -0,0 +1,295 @@
+from torch.distributions import Categorical
+
+from typing import Dict, List, Tuple
+from collections import namedtuple
+
+from ..networks.a2c_fcn import *
+from ...task import SimulatedTask
+from ....types import TaskState, TaskType
+from .globals import *
+from .model import *
+from .env import *
+
+import torchviz
+import math
+
+MappingLogs = namedtuple("MappingLogs",
+                         ("state", "next_state"))
+
+class SimpleAgent(RLModel):
+
+    def __init__(self, rl_env: RLBaseEnvironment, load_best_model: int = 0,
+                 execution_mode: str = "testing", lr: float = 0.999):
+        self.num_actions = rl_env.get_out_dim()
+        # S, (S, A) value networks
+        self.network = A2CNetworkNoGCN(rl_env.get_state_dim(), rl_env.get_out_dim())
+        self.optimizer = optim.RMSprop(self.network.parameters(),
+                                       lr=0.0001)
+                                       #lr=0.0005)
+        self.execution_mode = execution_mode
+        self.is_loaded_model_best = load_best_model
+        self.fastest_execution_time = float("inf")
+        self.lr = lr
+        self.episode = 0
+        self.a2cnet_fname = "a2c_network.pt"
+        self.optimizer_fname = "optimizer.pt"
+        self.best_a2cnet_fname = "best_a2c_network.pt"
+        self.best_optimizer_fname = "best_optimizer.pt"
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        # Accumulated reward 
+        self.accumulated_reward = 0
+
+        # pi(state) -> action probabilities
+        self.pi = {}
+        # lambda to update a target policy
+        self.ld = 0.5
+
+        # Buffer (S, A, S') until the terminal state
+        self.tmp_curr_state = None
+        self.tmp_next_state = None
+        self.tmp_action = None
+        self.logs: List[MappingLogs] = []
+
+        if not self.is_training_mode():
+            print("Load model..\n")
+            if self.is_loaded_model_best == 1:
+                self.load_best_model()
+            else:
+                self.load_model()
+
+    def select_device(self, state: torch.tensor, oracle):
+        """
+        Select a device from pi.
+        If a specified state has not been visited, select a device from a neural
+        network.
+        """
+
+        state_tuple = tuple(state.tolist())
+        with torch.no_grad():
+            if state_tuple in self.pi:
```

Comment:
I don't understand this

Reply:
Why does the pi(S) -> action need to be stored for all states?

Reply:
This is also the thing that I confused (So I tried to ask this question to you in the slack).
pi(S) is a mapping bewteen S to action probs, correct(so, like a dictionary)? Then how can I avoid to store states?
or do you mean that it should be a function with an equation? what is that equation?

Reply:
also this pi(S) should be updated for each iteration (so, at the end of an iteration k, pi_{k-1}(S), but this will be updated based on oracle and f(theta) at an iteration k, to pi_{k}(S)

Reply:
You only need to store the pi associated with a given action choice, iiuc

`
mapping_function():
        # get current state S
        # compute pi, p, v 
        # use pi to decide mapping a
        buffer.add( (S, pi, p, v, a) ) 
        
optimize_network():
          # use buffer to compute loss 
          # backprop & step
          # clear buffer
`

Where:
 S is a feature vector
 pi is a vector of length(# devices)
 p is a vector of length(# devices)
 v is a scalar
 a is a scalar

Reply:
I don't get it. Sorry. what do you mean by "compute pi"? 
Then, do you mean that we have an equation to compute pi?
What is that?
lambda*f + (lambda-1)*oracle is an equation to "update" pi, isn't it?

Reply:
Given a state S and a network function f  `lambda*f(S) + (lambda-1)*oracle(S)` is a function that computes pi. 
I am not sure what you mean by update.

Reply:
ok. that's my misunderstanding. in the jamboard, this was introduced as "update" function. in alphago, it has an update function from bottom to top with its simulated `v`. So there is no update but only decision from oracle. This is good and now I think I understood it. My current code is then incorrect. Its updating pi function also should be removed.